### PR TITLE
VisualStateTrigger

### LINKED ISSUE #7632: [Enhancement] GoToVisualState for XAML
## Summary
Allow to use VisualState.GoTo in triggers

## API Changes
``` xaml
<DataTrigger.EnterActions>
    <GoToVisualState State="Focus" />
</DataTrigger.EnterActions>
```

### PR DESCRIPTION
**Pull request by KSemenenko**
### Description of Change ###
I think that there is not enough trigger that would allow using VisalStateManager from XAML.
I added a trigger that will allow this to be done.

### Issues Resolved ### 

- fixes #7632

### API Changes ###
``` xaml
<Entry.Triggers>
    <EventTrigger Event="Focused">
        <VisualStateTrigger State="Focused"/>
    </EventTrigger>
    <EventTrigger Event="TextChanged">
         <VisualStateTrigger State="Text" Target="{x:Reference BoxForState}" />
    </EventTrigger>
    <DataTrigger.EnterActions>
        <VisualStateTrigger State="Focus" />
    </DataTrigger.EnterActions>
</Entry.Triggers>
````

**GoToState** - the textual name of the state
**Target** - an optional parameter, a link to a specific object.

### Platforms Affected ### 

- Core/XAML (all platforms)

### Testing Procedure ###
- Find VisualStateManager Gallery
- Select "Visual States change by trigger"
- Select Entry
- Start typeing 

### PR Checklist ###

- [x] Targets the correct branch
- [x] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by KSemenenko** [comment]
@rmarinho can you run tests please?

**Comment by KSemenenko** [comment]
@StephaneDelcroix do you think this VisualStateTrigger is a good name for a trigger?

**Comment by KSemenenko** [comment]
@PureWeen @kingces95 this pr is ready to review

**Comment by KSemenenko** [comment]
nice addition option #7892

**Comment by KSemenenko** [comment]
@kingces95 @PureWeen @samhouts Maybe you have already watched this PR and have any comments?

**Comment by KSemenenko** [comment]
@kingces95 @PureWeen I updated the branch from the last develop

**Comment by KSemenenko** [comment]
@samhouts @kingces95 @PureWeen 
any updates?

File: Xamarin.Forms.Core/Interactivity/VisualStateTrigger.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using Xamarin.Forms.Xaml;
+
+namespace Xamarin.Forms
+{
+	public class VisualStateTrigger : TriggerAction<VisualElement>
+	{
+		public string GoToState { get; set; }
```
Comment (hartez) [inline Xamarin.Forms.Core/Interactivity/VisualStateTrigger.cs]: I think this property should be named `State`.

**Comment by rmarinho** [comment]
I wonder if this would be better is to add Triggers to a VisualState , and inside there we an have a Collection of Triggers that start that state, like in UWP. 

https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.visualstate.statetriggers#Windows_UI_Xaml_VisualState_StateTriggers

File: Xamarin.Forms.Core/Interactivity/VisualStateTrigger.cs (hunk 1 above)
Comment (KSemenenko) [reply Xamarin.Forms.Core/Interactivity/VisualStateTrigger.cs]: done

**Comment by KSemenenko** [comment]
@rmarinho So that we can use a trigger in any component.
As in my demo example, I can change the state of any component based on a trigger.

**Comment by PureWeen** [comment]
@KSemenenko I don't quite follow this thought

>  So that we can use a trigger in any component.

@jsuarezruiz has a PR that we're going to merge soon here with State based triggers that I think can achieve everything you've done here

https://github.com/xamarin/Xamarin.Forms/pull/9422

These triggers can be applied inside the VisualStateManager of any component and then transition its state based on the conditions of the State Trigger

**Comment by KSemenenko** [comment]
@PureWeen I mean, I want to be able to use triggers in XAML. Without using C# code.
For example, when some event happens and I want to change the state of control.
So I didn’t put it right, meaning that with the help of this trigger I can change the state of any control on the page without C#. and togeter with #7892 we can use VSM from ViewModel. Which will be very convenient, as for me. :)

**Comment by jsuarezruiz** [comment]
@KSemenenko  What do you think if we move it to the Toolkit?

**Comment by KSemenenko** [comment]
I love this idea :) @jsuarezruiz

**Comment by jfversluis** [comment]
This will not happen anymore for Xamarin.Forms. If this is still important to you, make sure to check the [.NET MAUI repo](https://github.com/dotnet/maui) and see if it's already on the roadmap or open an issue with a detailed feature request.

We appreciate the work that has gone into this PR although it won't be merged. Thank you so much for your precious time that you have donated to us. Hopefully it was still a fun and a great learning experience for you :)

I see there is a PR for this on the Toolkit, hopefully it will live a happy life there!

**Pull request closed without merging**